src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java

```diff
@@ -371,5 +372,56 @@ public CloudSdk build() {
           startListeners, runDevAppServerWaitSeconds);
     }
 
+    /**
+     * Attempt to find the Google Cloud SDK in various places.
+     * 
+     * @return the path to the root of the Google Cloud SDK
+     * @throws AppEngineException if not found
+     */
+    @Nonnull
+    private Path discoverSdkPath() {
+      for (CloudSdkResolver resolver : getResolvers()) {
+        try {
+          Path discoveredSdkPath = resolver.getCloudSdkPath();
+          if (discoveredSdkPath != null) {
+            return discoveredSdkPath;
+          }
+        } catch (Exception ex) {
+          logger.log(Level.SEVERE, resolver.getClass().getName()
+              + ": exception thrown when searching for Google Cloud SDK", ex);
+        }
+      }
+      throw new AppEngineException("Google Cloud SDK path was not provided and could not be"
+          + " found in any known install locations.");
+    }
+
+    /**
+     * Return the configured SDK resolvers.
+     */
+    @VisibleForTesting
+    public List<CloudSdkResolver> getResolvers() {
+      if (this.resolvers != null) {
+        return resolvers;
+      }
+      // Explicitly specify classloader rather than use the TCCL
+      ServiceLoader<CloudSdkResolver> services =
+          ServiceLoader.load(CloudSdkResolver.class, getClass().getClassLoader());
+      List<CloudSdkResolver> resolvers = Lists.newArrayList(services);
+      // Explicitly add the PATH-based resolver last as ServiceLoader
+      // does not provide any ordering
+      resolvers.add(new PathResolver());
+      return resolvers;
+    }
+
+    /*
+     * Set the list of path resolvers to locate the Google Cloud SDK. Intended for tests to
+     * precisely control where the SDK may be found.
+     */
+    @VisibleForTesting
+    public Builder resolvers(List<CloudSdkResolver> resolvers) {
```

Comment:
I think it should be a `List` — there should be an order.

The real value here is the use of `ServiceLocator` such that we disconnect the consumer of the `CloudSdk` from the providers who know how to best find the SDK.  The only real users of `resolvers()` will be tests.  After all, if the creator of the `CloudSdk.Builder` wants to add a resolver, then they know (or have the resolver instance to compute) the path and can set `sdkPath` explicitly.